// kick-api/index.js
const https = require('https')

function lookupUser(username) {
    const data = {
        username: username
    };

    const postData = JSON.stringify(data);

    const options = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': postData.length
      }
    };
  
    return new Promise((resolve, reject) => {
      const req = https.request("https://api.kickalerts.live/api/v1/proxy/lookup", options, (res) => {
        let body = '';
        res.on('data', (chunk) => {
          body += chunk;
        });
        res.on('end', () => {
          resolve(JSON.parse(body));
        });
      });
  
      req.on('error', (err) => {
        reject(err);
      });
  
      req.write(postData);
      req.end();
    });
}

function followCount(username){
    const data = {
        username: username
    };

    const postData = JSON.stringify(data);

    const options = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': postData.length
      }
    };
  
    return new Promise((resolve, reject) => {
      const req = https.request("https://api.kickalerts.live/api/v1/proxy/followcount", options, (res) => {
        let body = '';
        res.on('data', (chunk) => {
          body += chunk;
        });
        res.on('end', () => {
          resolve(JSON.parse(body));
        });
      });
  
      req.on('error', (err) => {
        reject(err);
      });
  
      req.write(postData);
      req.end();
    });
}


function viewCount(username){
    const data = {
        username: username
    };

    const postData = JSON.stringify(data);

    const options = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': postData.length
      }
    };
  
    return new Promise((resolve, reject) => {
      const req = https.request("https://api.kickalerts.live/api/v1/proxy/viewcount", options, (res) => {
        let body = '';
        res.on('data', (chunk) => {
          body += chunk;
        });
        res.on('end', () => {
          resolve(JSON.parse(body));
        });
      });
  
      req.on('error', (err) => {
        reject(err);
      });
  
      req.write(postData);
      req.end();
    });
}

module.exports = {
  lookupUser,
  followCount,
  viewCount
};
